packages/platforms/accton/x86-64/x86-64-accton-as5912-54xk/onlp/builds/src/module/src/sfpi.c

```diff
@@ -0,0 +1,325 @@
+/************************************************************
+ * <bsn.cl fy=2014 v=onl>
+ *
+ *           Copyright 2014 Big Switch Networks, Inc.
+ *           Copyright 2016 Accton Technology Corporation.
+ *
+ * Licensed under the Eclipse Public License, Version 1.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ *        http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the
+ * License.
+ *
+ * </bsn.cl>
+ ************************************************************
+ *
+ *
+ *
+ ***********************************************************/
+#include <onlp/platformi/sfpi.h>
+#include <onlplib/file.h>
+#include "platform_lib.h"
+
+#include "x86_64_accton_as5912_54xk_log.h"
+
+#define NUM_OF_SFP_PORT 54
+#define MAX_SFP_PATH 	64
+static char sfp_node_path[MAX_SFP_PATH] = {0};
+#define FRONT_PORT_BUS_INDEX(port) (port+26)
+
+static char*
+sfp_get_port_path_addr(int port, int addr, char *node_name)
+{
+    sprintf(sfp_node_path, "/sys/bus/i2c/devices/%d-00%d/%s",
+                           FRONT_PORT_BUS_INDEX(port), addr, node_name);
+    return sfp_node_path;
+}
+
+static char*
+sfp_get_port_path(int port, char *node_name)
+{
+    return sfp_get_port_path_addr(port, 50, node_name);
+}
+
+/************************************************************
+ *
+ * SFPI Entry Points
+ *
+ ***********************************************************/
+int
+onlp_sfpi_init(void)
+{
+    /* Called at initialization time */    
+    return ONLP_STATUS_OK;
+}
+
+int
+onlp_sfpi_bitmap_get(onlp_sfp_bitmap_t* bmap)
+{
+    /*
+     * Ports {0, 54}
+     */
+    int p;
+    AIM_BITMAP_INIT(bmap, 64);
+    AIM_BITMAP_CLR_ALL(bmap);
+    
+    for(p = 0; p < NUM_OF_SFP_PORT; p++) {
+        AIM_BITMAP_SET(bmap, p);
+    }
+
+    return ONLP_STATUS_OK;
+}
+
+int
+onlp_sfpi_is_present(int port)
+{
+    /*
+     * Return 1 if present.
+     * Return 0 if not present.
+     * Return < 0 if error.
+     */
+    int present;
+    char *path = sfp_get_port_path(port, "sfp_is_present");
+
+    if (onlp_file_read_int(&present, path) < 0) {
```

Comment:
The ONLP File APIs allow you to specify a format string directly. You can replace the construction of "path" here with a direct call.